openerp/addons/base/ir/ir_attachment.py

```diff
@@ -160,23 +170,47 @@ def _file_write(self, cr, uid, value, checksum):
             try:
                 with open(full_path, 'wb') as fp:
                     fp.write(bin_value)
+                # add fname in checklist, in case the current transaction aborts
+                self._check(cr, uid, fname)
             except IOError:
                 _logger.info("_file_write writing %s", full_path, exc_info=True)
         return fname
 
     def _file_delete(self, cr, uid, fname):
-        # using SQL to include files hidden through unlink or due to record rules
-        cr.execute("SELECT COUNT(*) FROM ir_attachment WHERE store_fname = %s", (fname,))
-        count = cr.fetchone()[0]
-        full_path = self._full_path(cr, uid, fname)
-        if not count and os.path.exists(full_path):
-            try:
-                os.unlink(full_path)
-            except OSError:
-                _logger.info("_file_delete could not unlink %s", full_path, exc_info=True)
-            except IOError:
-                # Harmless and needed for race conditions
-                _logger.info("_file_delete could not unlink %s", full_path, exc_info=True)
+        # simply add fname in checklist
+        self._check(cr, uid, fname)
+
+    def _file_gc(self, cr, uid, context=None):
+        """ Trigger garbage collection of the filestore. """
+        location = self._storage(cr, uid, context)
+        if location != 'file':
+            return
+        # prevent all concurrent updates on ir_attachment while collecting!
+        cr.execute("LOCK ir_attachment IN SHARE MODE")
+        # retrieve the file names in the checklist, and process them
+        checked = 0
+        removed = 0
+        for dirpath, _, filenames in os.walk(self._full_path(cr, uid, "check")):
+            dirname = os.path.basename(dirpath)
+            for filename in filenames:
+                checked += 1
+                fname = "%s/%s" % (dirname, filename)
+                cr.execute("SELECT COUNT(*) FROM ir_attachment WHERE store_fname=%s", [fname])
```

Comment:
It might be possible to do that with a single query with a `SELECT … EXCEPT`, or maybe by loading filenames to check in a temp table and performing the relevant join.